### ISSUE TITLE
[BUG] S22i connects and reports info to QZ, but Auto-Resistance is not being controlled by QZ App

### ISSUE THREAD

**Issue by bkeown64** [cagnulein/qdomyos-zwift#2909]
When connecting QZ to the NordicTrack S22i, the QZ app reports the metrics, but when running a Peloton class, the auto-resistance is not changing when instructors calls out new resistance in class.  

I apologize for the blurry pictures.  Hard to take a photo when trying to get the S22i console running, getting an Android tablet to run the peloton class, my iPhone to run the QZ app, get the pedals going on the bike and taking pictures with an iPad!

The only other glitch I am getting is that the cadence seems to drop off periodically, every few seconds, and then re-connects.  It looks like everything is close to working except the autoresistance changing.  I am thinking it may just be a simple setting issue.  I have attached a debug file as well.  

Thank you for all you do. 

![IMG_0003](https://github.com/user-attachments/assets/4f261445-7cfd-4c09-80eb-0b4455fc2028)

![IMG_0005](https://github.com/user-attachments/assets/6c506d4b-21e7-44ea-9f5f-793d136d822a)

**Screenshots**

Bike is Nordictrack S22i
QZ running on iPhone 15 Pro Max using OS 18.1.1
Android is the Onn Tablet Pro, 11 inch, 2024 edition running Android 14.

[debug-Mon_Dec_16_09_10_51_2024.log](https://github.com/user-attachments/files/18151886/debug-Mon_Dec_16_09_10_51_2024.log)

**Comment by cagnulein** [comment]
did you select the s22i in the qz companion app?
there are 2 models of it, you can try both. restart the bike everytime you changed the device in the companion

**Comment by bkeown64** [comment]
I did select the s22i in the companion app, but didn't notice that there
are two listed.  I'll try the second one and report back.

**Comment by bkeown64** [comment]
Yes I see the bike resistance changing manually

**Comment by cagnulein** [comment]
Don't you see a touch reaction on the bike's console? Qz faking a touch
event on the bike's tablet when your press the plus button on the
inclination and/or resistance

If so this is the main issue

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 16 dic 2024 alle 17:23 bkeown64 ***@***.***> ha
scritto:

**Comment by bkeown64** [comment]
Sorry but not sure what you mean by touch reaction on the bike's console.
I assume you mean that if the resistance is attempted to be changed on the
QZ app, I should see the corresponding change on the bike's console.  If
that is what you mean, then no...when touching the + or minus on the
resistance tile in QZ, the resistance does not change on the app or on the
bike.  However, if I touch the + or _ buttons on the resistance handle on
the bike, the bike will adjust resistance and the new resistance will show
up on the QZ app.

**Comment by bkeown64** [comment]
No sir.  Do you have a picture of it?  You aren't talking about the speed
and resistance sliding scales on the screen, I take it (which they wont
show unless I turn them on in the menu)

**Comment by bkeown64** [comment]
As an FYI, I tried to install QZ with the https://tinyurl.com/qz-companion
link.  The file downloads but I get an error message stating "Could not
load file".    I had to download a version from a mirror site (
https://android.izzysoft.de/repo/apk/org.cagnulein.qzcompanionnordictracktreadmill)
and it installed.  I tried using both s22i profiles in the companion app
but have the same problem with either selected.

**Comment by cagnulein** [comment]
Yes I was talking about the inclination and resistance slider.
Anyway I guess I understood the issue, now it's late here, I will be back
tomorrow. I guess we need to use the accessibility engine that I developed
the last month


Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 16 dic 2024 alle 20:34 bkeown64 ***@***.***> ha
scritto:

**Comment by bkeown64** [comment]
Thank you for your help!

**Comment by cagnulein** [comment]
Could you try to enable the x22i no adb profile in the companion and
restart?

You should then see random touchscreen event if you press the plus buttons
in qz

If so I just need to use add the same behavior

Let me know

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 16 dic 2024 alle 20:38 bkeown64 ***@***.***> ha
scritto:

**Comment by bkeown64** [comment]
I did and still no luck.  This time I enabled the sliders on the bike
display.  Again, the app is reporting everything accurately and even within
the app, it provides the Peloton-prescribed resistance and target
resistance, plus reports the bike's set resistance when I manually change
it on the bike.  After restarting the bike and app again, I did notice that
within the QZ app I could now manually change the resistance, but the
resistance in the app quickly changes back to the Peloton instructor's
resistance - but no changes occur on the bike itself.

**Comment by cagnulein** [comment]
Did you enable accessibility permission in the android settings on the
bike's tablet for the qz companion?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 16 dic 2024 alle 21:14 bkeown64 ***@***.***> ha
scritto:

**Comment by bkeown64** [comment]
Yes I did

**Comment by cagnulein** [comment]
Does  adb work from your pc /Mac?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 16 dic 2024 alle 21:38 bkeown64 ***@***.***> ha
scritto:

**Comment by bkeown64** [comment]
I have not had any experience doing that, but I will look into it.  I
dont have a PC and I understand it is a bit more difficult to do with a
mac.

**Comment by bkeown64** [comment]
Good morning, Roberto.  Unfortunately the steps for installing on a Mac
aren't exactly the same.  Following the directions in the Readme.md file,
you cannot right click the MacOS install file from github to run the
application from terminal.  I also downloaded the RAW file and attempted to
run it from the file I saved on my desktop with no luck either.

**Comment by cagnulein** [comment]
you can download this https://github.com/cagnulein/QZCompanionNordictrackTreadmill/archive/refs/heads/master.zip

and then open a terminal there and run

[QZ-Companion-MacOS](https://github.com/cagnulein/QZCompanionNordictrackTreadmill/blob/master/InstallPackage/QZ-Companion-MacOS)

you would need probably to do a 

chmod a+x QZ-Companion-MacOS

and then

./QZ-Companion-MacOS

if everything works fine then we should have access to adb and try to change the input touchscreen via adb

let me know

**Comment by bkeown64** [comment]
Thanks, Robert.  It installed easily via the Zip package extraction, even
without the chmod command.  When testing out with a Peloton class, it
immediately began changing elevation in a Peloton climb class, but wouldn't
change the resistance from the instructor callouts.  When testing with
Zwift, it did change the elevations and I assume that resistance must be
manually controlled (via the handlebars resistance + and - key, which is
fine because it emulates bike gears in this mode).

Sounds like I am very close to getting this working perfectly.  One
question - in your Readme file, I see that under Experimental Settings,
Zwift Force Resistance is unchecked.  I left mine checked on in the default
position.  Should I uncheck it for Peloton and check it for Zwift?

**Comment by cagnulein** [comment]
> When testing out with a Peloton class, it immediately began changing elevation in a Peloton climb class, but wouldn't change the resistance from the instructor callouts.

wow this is amazing! do you have both logs so i can compare them?



> One question - in your Readme file, I see that under Experimental Settings, Zwift Force Resistance is unchecked. I left mine checked on in the default position. Should I uncheck it for Peloton and check it for Zwift?

it must be enabled, you did right

**Comment by bkeown64** [comment]
Sorry, the debug files are for some reason too large to send
conventionally.  Yes, Zwift Force Resistance was enabled for both classes.

**Comment by bkeown64** [comment]
I have a different profile setup for Peloton classes.  Hopefully the
archived files I just sent show that

**Comment by cagnulein** [comment]
nope unfortunately, that one was a zwift class

**Comment by cagnulein** [comment]
i don't see it @bkeown64 you have to upload it to the github web interface
let me know

**Comment by bkeown64** [comment]
[debug-Tue_Dec_17_08_53_08_2024.log.zip](https://github.com/user-attachments/files/18168954/debug-Tue_Dec_17_08_53_08_2024.log.zip)

**Comment by bkeown64** [comment]
Ran another 3 minutes of a climb class and have attached the debug file for it.  Yes - Peloton credentials are entered successfully.
[debug-Tue_Dec_17_12_49_21_2024.log.zip](https://github.com/user-attachments/files/18169623/debug-Tue_Dec_17_12_49_21_2024.log.zip)

**Comment by bkeown64** [comment]
What is the issue?

**Comment by cagnulein** [comment]
you should tell me :) in the previous logs the peloton credentials was wrong, now they are correct :)

**Comment by bkeown64** [comment]
lol…sorry I must have sent the wrong log.

**Comment by cagnulein** [comment]
Ok but so could you generate the log with the peloton auto resistance issue?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 17 dic 2024 alle 19:45 bkeown64 ***@***.***> ha
scritto:

**Comment by bkeown64** [comment]
That was a debug file with the auto-resistance issues.  As I said before, everything - including resistance - changes in the app per the instructor callouts, but the auto-resistance on the bike is not changing, only the incline.  I have the older S22i bike and I am currently using the one that says just s22i in the app (not the one with the model number listed after it).  However, I had the same issue with either one selected.

**Comment by cagnulein** [comment]
Got it will check the log tomorrow

**Comment by bkeown64** [comment]
Ok, thanks.

**Comment by cagnulein** [comment]
@bkeown64 ok i got it. in the companion, for your bike, i'm handling only the inclination because, the original user that sent me the information, told me that changing the inclination also have affect on the resistance of the bike. Isn't the case? I mean, without looking at the resistance value on the slider, if you change only the inclination from the ifit app, don't you feel also the resistance increase on the legs?

**Comment by bkeown64** [comment]
Theoretically that is true, but I doubt that resistance number is
calculable for Peloton. If your programming could provide for Peloton
classes the standard resistance changes only like your app does for most of
the Proform bikes and Echelon bikes and then leave the resistance provided
through inclination for Zwift, ROUVY, MyWhoosh, etc. then you have the best
of both worlds.

**Comment by cagnulein** [comment]
the conversion should be this

![image](https://github.com/user-attachments/assets/f0d81a15-0f8e-4410-ba7c-19ebea25338b)

we can scale this for the max inclination of the bike and gain the same thing.

what do you think? this is easiest way to do it if you can confirm me that with the max inclination and resistance 1 and the max resistance and inclination 0 you have the same effort.

otherwise i have to implement also the resistance thing for the s22i (it could be more a pain :) but it's doable)

**Comment by bkeown64** [comment]
I absolutely can do that for you.  The bike incline is definitely a
paradigm shift in managing resistance and I do understand that it  probably
will be more effort programming that in.  Wisely, save that effort for when
the demand is there for it.

I'll shoot that info to you here shortly.  So If I understand correctly,
you want me to hop on the bike and report to you the watt effort produced
for when the bike is at incline 0 and incline level 1?  Do you want that at
a standard 80 cadence or an 80 and 100 cadence?

**Comment by cagnulein** [comment]
> So If I understand correctly, you want me to hop on the bike and report to you the watt effort produced for when the bike is at incline 0 and incline level 1? Do you want that at a standard 80 cadence or an 80 and 100 cadence?

not exactly i would need this:

- wattage at cadence 80 with inclination at the max and resistance to the min
- wattage at cadence 80 with inclination at the min and resistance to the max

I'm expecting that should be more or less the same. If not we need just to find a scale for this and this would be easy to setup.

**Comment by bkeown64** [comment]
Got it,  I will have it in a few, however, I hope I can pedal 80 cadence at
max level!!!!

**Comment by bkeown64** [comment]
Wow….. let’s just say yes, they are equal. There was no way I could get
even close to even 30 cadence at either level.  I was very briefly (less
than a second)at 80 cadence at both scenarios and saw about 600 watts on
resistance max, incline 0 and around 500 watts at incline 20, resistance 0

**Comment by cagnulein** [comment]
ok that's perfect! let me finalize the code for this so

**Comment by bkeown64** [comment]
Just to add, if you alert those wanting to use these Nordictrack bikes for
Peloton that the resistance is calculated on wattage produced at 80 cadence
for a specific incline and at 0 resistance, then this should be fine as
long as the user understands this.

For over 6 months now, I have used your app daily on several bikes,
including an Echelon Connect and an EX-5, a Proform Sport CX bike, and a
Schwinn IC4, and it caught me by surprise how inclination was being used to
manage resistance for Peloton classes and something I didn't expect.  The
Nordictrack bikes are definitely a breed of their own.

**Comment by bkeown64** [comment]
Crap!  Didnt install the new version yet,  Let me try again.  Bear with me!

**Comment by bkeown64** [comment]
[debug-Wed_Dec_18_12_25_46_2024.log](https://github.com/user-attachments/files/18186811/debug-Wed_Dec_18_12_25_46_2024.log)

Bit of a regression here.  The Peloton app connected and the QZ App is receiving data from the bike and the Peloton class, but is not changing the inclination at all.  Incline was at 0 and Resistance stayed at 1 as well for the three and half minutes I took the class.

**Comment by cagnulein** [comment]
Argh i will check it, thanks

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 18 dic 2024 alle 18:33 bkeown64 ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
Actually in the log i can see Qz asking to change the  inclination  to 5
and then to 8.
Did you try to restart the bike?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 18 dic 2024 alle 18:37 cagnulein ***@***.***> ha
scritto:

**Comment by bkeown64** [comment]
I didn’t power the bike down to restart it.  I did go into privileged mode, rechecked s22i in the Companion app, and then restarted iFit

**Comment by cagnulein** [comment]
I don't know i mean from the qz point of view it's the same as before, just
sending different values. Maybe you can try to rollback to the previous
version via testflight to check if it's working?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 18 dic 2024 alle 18:56 bkeown64 ***@***.***> ha
scritto:

> I didn’t power the bike down to restart it. I did go into privileged mode,
> rechecked s22i in the Companion app, and then restarted iFit
>

**Comment by bkeown64** [comment]
Apparently in the companion app after restarting it also changed it the bike profile.  After changing that back to s22i bike it started changing resistance again.

**Comment by bkeown64** [comment]
Again, the biggest obstacle is getting the understand how resistance is calculated. In my opinion the only true way for this to work on a Peloton class correctly is to have it change resistance only.  It is going to be confusing for the end user to understand the app.

**Comment by cagnulein** [comment]
for sure i can add the resistance instead. the fact the color doesn't match it's a bug for sure.

I will try to do this in the xmas days. Thanks for your feedbacks

**Comment by bkeown64** [comment]
No hurry Roberto.  Thank you for looking into this as I know there are a lot of s22i bikes out there that would love your app.  I'll even write the wiki up for you once it gets working ideally.

**Comment by bkeown64** [comment]
Awesome, thank you so much!
